nemo_aligner/models/nlp/gpt/megatron_gpt_rs_actor.py

```diff
@@ -0,0 +1,337 @@
+
+                if parallel_state.is_pipeline_last_stage():
+                    required_keys.update(("tokens", "advantages", "mask", "prev_log_probs"))
+
+            batch = {key: val.cuda(non_blocking=True) if key in required_keys else None for key, val in batch.items()}
+
+            parallel_logits = model(batch["tokens"], batch["position_ids"], batch["attention_mask"], labels=None,)
+
+            def loss_func(parallel_logits):
+                mask = batch["mask"]
+                tokens = batch["tokens"]
+
+                curr_log_probs = from_parallel_logits_to_logprobs(vocab_parallel_logits=parallel_logits, target=tokens)
+                loss = -1 * masked_mean(curr_log_probs, mask)  # Loss is mean logits on response tokens
+
+                reduced_actor_loss = average_losses_across_data_parallel_group([loss])
+                return (
+                    loss,
+                    {"loss": reduced_actor_loss,},
+                )
+
+            return parallel_logits, loss_func
+
+        return fwd_output_and_loss_func
+
+    def prepare_for_training(self):
+        configure_batch_sizes(
+            mbs=self.cfg.micro_batch_size,
+            gbs=self.cfg.global_batch_size,
+            dp=parallel_state.get_data_parallel_world_size(),
+        )
+        self.onload_adam_states()
+
+    def prepare_for_training_step(self):
+        # custom trainers will always zero grad for us
+        prepare_for_training_step(self, zero_grad=False)
+
+    def get_loss_and_metrics(self, batch, forward_only):
+        sequence_length = batch["response_tokens"].size(-1)
+
+        data_iter = get_iterator_k_split(batch, get_num_microbatches())
+        set_sync_funcs(self, forward_only)
+        fwd_bwd_function = get_forward_backward_func()
+
+        losses_reduced_per_micro_batch = fwd_bwd_function(
+            forward_step_func=self.get_actor_forward_output_and_loss_func(),
+            data_iterator=data_iter,
+            model=self.model,
+            num_microbatches=get_num_microbatches(),
+            forward_only=forward_only,
+            seq_length=sequence_length,
+            micro_batch_size=self.cfg.micro_batch_size,
+        )
+
+        metrics = {}
+
+        for key in ["loss"]:
+            if losses_reduced_per_micro_batch:
+                metric_mean = torch.stack(
+                    [loss_reduced[key] for loss_reduced in losses_reduced_per_micro_batch]
+                ).mean()
+            else:
+                metric_mean = torch.tensor(0.0, device=torch.cuda.current_device())
+
+            torch.distributed.broadcast(metric_mean, get_last_rank())
+
+            metrics[key] = metric_mean.cpu().item()
+
+        return metrics["loss"], metrics
+
+    def finish_training_step(self):
+        grad_reductions(self)
+
+    def finish_training(self):
+        """no need to offload adam states here
+        """
+
+    # inference calls
+    def get_logprob_output_only_func(self, inference_only=True):
```

Comment:
« O: » I suspect there are a bunch of methods (this one + more) that we don’t need anymore

Reply:
I have cleaned up the code and removed all the init policy stuff we are not using.

Reply:
Sounds good, I believe the only leftover remaining to remove is:
```python
    init_policy_state_dict = None
    ptl_model.init_policy_state_dict = init_policy_state_dict
```

Reply:
Done.